content/developer/tutorials/server_framework_101/06_security.rst

```diff
@@ -348,10 +348,375 @@ access to property records.
+  selected by the current user in the company switcher.
+
+.. example::
+   In the example below, we extend the product and product category models to support multi-company,
+   and define record rules to ensure proper data isolation between companies.
+
+   .. code-block:: python
+
+      class Product(models.Model):
+          _name = 'product'
+          _check_company_auto = True
+
+          company_id = fields.Many2one(string="Company", comodel_name='res.company')
+          price = fields.Float(string="Sales Price", required=True, default=100)
+          cost = fields.Float(string="Manufacturing Cost", company_dependent=True)
+          margin = fields.Float(
+              string="Profit Margin", compute='_compute_margin', inverse='_inverse_margin'
+          )
+          category_id = fields.Many2one(
+              string="Category",
+              comodel_name='product.category',
+              ondelete='restrict',
+              required=True,
+              default=lambda self: self.env.ref('product.category_apparel'),
+              check_company=True,
+          )
+
+          @api.depends('price', 'cost', 'company_id')
+          @api.depends_context('company')
+          def _compute_margin(self):
+              for product in self:
+                  product.margin = product.price - product.with_company(product.company_id).cost
+
+      class ProductCategory(models.Model):
+          _name = 'product.category'
+          _check_company_auto = True
+
+          company_id = fields.Many2one(
+              string="Company",
+              comodel_name='res.company',
+              required=True,
+              default=lambda self: self.env.company.id,
+          )
+          product_ids = fields.One2many(
+              string="Products", comodel_name='product', inverse_name='category_id', check_company=True
+          )
+
+   .. code-block:: xml
+
+      <record id="product.product_category_company_rule" model="ir.rule">
+          <field name="name">Product: Access product categories in own companies only</field>
+          <field name="model_id" ref="product.model_product_category"/>
+          <field name="domain_force">[('company_id', 'parent_of', company_ids)]</field>
+      </record>
+
+      <record id="product.product_company_rule" model="ir.rule">
+          <field name="name">Product: Access products in own companies only</field>
+          <field name="model_id" ref="product.model_product"/>
+          <field name="domain_force">['|', ('company_id', '=', False), ('company_id', 'parent_of', company_ids)]</field>
+      </record>
+
+   .. note::
+      - A `company_id` field is added to the `product` and `product.category` models, allowing them
+        to be company-specific.
+      - The `company_id` field is optional on the `product` model, allowing products to be shared
+        between companies. It is however required for the `product.category` model, making
+        categories company-specific.
```

Comment:
Considering that the category is required on the product, it's strange for a shared product to belong to a category belonging to a specific company (unless you make the field company_dependent).  Shouldn't it be the opposite ? :thinking:

Reply:
Yeah, it should, don't remember why I went in that direction... maybe to keep the example ~~simple~~ not too complex.

Reply:
Ok I remember, it was to use the existing fields of product as examples for company-dependency and computes, without inventing new fields on product.category just for the sake of the example.